### PR TITLE
Post: Duplication, you're welcome

### PR DESCRIPTION
**Pull request by rachelcarmena**

### CODE REVIEW

File: _posts/2018-02-27-duplication-you-are-welcome.md [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+image:
+   src: /assets/custom/img/blog/2018-02-27-duplication-you-are-welcome/duplication-header.jpg
+tags:
+- quality
+- duplication
+- metrics
+
+---
+
+There are models to describe the process of learning as <a href="http://wiki.c2.com/?DreyfusModel">Dreyfus model</a> or <a href="http://wiki.c2.com/?ShuHaRi" target="_blank">Shuhari</a>, but I'd explain another process when learning a new technical skill. It would be composed by 2 phases:
```
Comment (richardjwild) [inline _posts/2018-02-27-duplication-you-are-welcome.md]: composed _of_

File: _posts/2018-02-27-duplication-you-are-welcome.md [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+tags:
+- quality
+- duplication
+- metrics
+
+---
+
+There are models to describe the process of learning as <a href="http://wiki.c2.com/?DreyfusModel">Dreyfus model</a> or <a href="http://wiki.c2.com/?ShuHaRi" target="_blank">Shuhari</a>, but I'd explain another process when learning a new technical skill. It would be composed by 2 phases: 
+
+1. **Fascination**. When we learn a new concept that sounds really good to solve our problems and we apply it in any place we have an apportuninity. And we are so convinced about its advantages that it seems we have a radar to detect a place in which we can apply it. It's like the first stage at Shuhari, but we are absolutely fascinated and we don't believe in _another world_.
```
Comment (richardjwild) [inline _posts/2018-02-27-duplication-you-are-welcome.md]: 1. _This is_ when we learn a new concept
2. Place a comma after problems
3. Lose "and" at the beginning of the second sentence, i.e. "We are so convinced..."

File: _posts/2018-02-27-duplication-you-are-welcome.md [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+
+2. **Questioning**. When we have assimilated that concept and we are able to decide where it's convenient to implement it. We have learnt from the previous stage and now we are moderate and reasonable about our decisions.
```
Comment (richardjwild) [inline _posts/2018-02-27-duplication-you-are-welcome.md]: Again, begin with "This is when..." or you could use "Questioning is when..." for variety if you like.

File: _posts/2018-02-27-duplication-you-are-welcome.md [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+A few weeks ago, I helped to separate two different modules. They were together, because they had some similar domain objects, although they didn't use the same properties. It was really difficult to know which properties and methods needed each module.
+
+What's the reason of those extracted methods or that reuse?:
+
+`Alerts: duplicated lines (%) > 10%`
+
+One of the metrics we can find in static code analysis tools is the **percentage of duplicated lines**. Managers need metrics and this one is usually included as a guarantee of **quality code**. However, striving for a certain number of duplicated lines in a board can cause these consequences:
+
+* Over-generalization
+* Over-anticipate abstractions
```
Comment (richardjwild) [inline _posts/2018-02-27-duplication-you-are-welcome.md]: Either "over-anticipated abstractions" or "over-anticipating abstractions"

File: _posts/2018-02-27-duplication-you-are-welcome.md [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+
+This principle is talking about **knowledge**, not duplicated lines. Two pieces of code could be apparently similar, but have completely different reasons to change, because they represent different pieces of knowledge. Therefore, removing that duplication can muddle the logic and make any future change worse.
+
+That's the reason why I usually highlight **knowledge** when explaining the third rule of <a href="http://wiki.c2.com/?XpSimplicityRules" target="_blank">simple design</a>:
+> No **knowledge** duplication
+
+To sum up, some recommendations:
+
+* Try to postpone your decisions about duplicated code, because it's a question of trade-offs
+* When you find two apparently similar pieces of source code, question if you find a reason to have to change them at the same time
```
Comment (richardjwild) [inline _posts/2018-02-27-duplication-you-are-welcome.md]: question _whether_

File: _posts/2018-02-27-duplication-you-are-welcome.md [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+* Use the metric about duplicated lines only as a **guidance**, but don't try to strive for a number in a board, compromising the understanding of source code
```
Comment (richardjwild) [inline _posts/2018-02-27-duplication-you-are-welcome.md]: "only as a guide" or "only as guidance"

File: _posts/2018-02-27-duplication-you-are-welcome.md (hunk 1 above)
Comment (richardjwild) [reply _posts/2018-02-27-duplication-you-are-welcome.md]: I'd _like to_ explain

Comment (richardjwild) [reply _posts/2018-02-27-duplication-you-are-welcome.md]: It _is_ composed

File: _posts/2018-02-27-duplication-you-are-welcome.md [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+tags:
+- quality
+- duplication
+- metrics
+
+---
+
+There are models to describe the process of learning as <a href="http://wiki.c2.com/?DreyfusModel">Dreyfus model</a> or <a href="http://wiki.c2.com/?ShuHaRi" target="_blank">Shuhari</a>, but I'd like to explain another process when learning a new technical skill. It is composed of 2 phases: 
+
+1. **Fascination**. This is when we learn a new concept that sounds really good to solve our problems, and we apply it in any place we have an apportuninity. We are so convinced about its advantages that it seems we have a radar to detect a place in which we can apply it. It's like the first stage at Shuhari, but we are absolutely fascinated and we don't believe in _another world_.
```
Comment (richardjwild) [inline _posts/2018-02-27-duplication-you-are-welcome.md]: apportuninity -> opportunity

**Comment by rachelcarmena** [comment]
Thanks @richardjwild !! :heart:

File: _posts/2018-02-27-duplication-you-are-welcome.md [hunk 8]
```diff
@@ -17,7 +17,7 @@ tags:
 
 There are models to describe the process of learning as <a href="http://wiki.c2.com/?DreyfusModel">Dreyfus model</a> or <a href="http://wiki.c2.com/?ShuHaRi" target="_blank">Shuhari</a>, but I'd like to explain another process when learning a new technical skill. It is composed of 2 phases: 
 
-1. **Fascination**. This is when we learn a new concept that sounds really good to solve our problems, and we apply it in any place we have an apportuninity. We are so convinced about its advantages that it seems we have a radar to detect a place in which we can apply it. It's like the first stage at Shuhari, but we are absolutely fascinated and we don't believe in _another world_.
+1. **Fascination**. This is when we learn a new concept that sounds really good to solve our problems, and we apply it in any place we have an apportunity. We are so convinced about its advantages that it seems we have a radar to detect a place in which we can apply it. It's like the first stage at Shuhari, but we are absolutely fascinated and we don't believe in _another world_.
```
Comment (richardjwild) [inline _posts/2018-02-27-duplication-you-are-welcome.md]: I'm sorry! It's still wrong! 😢 It's opportunity with an o.

**Comment by rachelcarmena** [comment]
Don't worry @richardjwild !! You have unlimited opportunities :wink: Thank you!!

**Pull request merged**